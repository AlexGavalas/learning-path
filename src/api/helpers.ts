import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
