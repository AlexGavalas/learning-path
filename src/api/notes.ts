import type { AstroComponentFactory } from 'astro/dist/runtime/server';
import { getEntry, render } from 'astro:content';
import flow from 'lodash/fp/flow';
import groupBy from 'lodash/fp/groupBy';
import mapValues from 'lodash/fp/mapValues';

import type { BlogArticleFrontmatter } from '~types/blog';
import type { Note, NoteFrontmatter } from '~types/notes';
import type { SummaryFrontmatter } from '~types/summaries';

import { getNoteMetadata, searchNotes } from './notes-db';

type SearchNotesRpcResponse = {
    title: string;
    line: string;
}[];

type Lines = Record<string, string[]>;

const groupByTitle = flow(
    groupBy<SearchNotesRpcResponse[number]>(({ title }) => title),
    mapValues((notes) => notes.map(({ line }) => line)),
);

export const fetchNotes = async (
    query: string,
): Promise<{
    lines: Lines;
    notes: Note[];
}> => {
    const allNotes = await getNoteMetadata();

    if (query.length > 0) {
        const rows = await searchNotes(query);

        const lines = groupByTitle(rows);

        const noteTitles = new Set(rows.map(({ title }) => title));

        const filteredNotes = allNotes.filter(({ title }) =>
            noteTitles.has(title),
        );

        return {
            lines,
            notes: filteredNotes,
        };
    }

    return {
        lines: {},
        notes: allNotes,
    };
};

export const getNoteData = async ({
    collection,
    slug,
}: {
    collection: 'blog' | 'notes' | 'summaries';
    slug: string;
}): Promise<{
    content: { Content: AstroComponentFactory };
    frontmatter: NoteFrontmatter | SummaryFrontmatter | BlogArticleFrontmatter;
} | null> => {
    const note = await getEntry(collection, slug);

    if (!note) {
        return null;
    }

    return {
        content: await render(note),
        frontmatter: note.data,
    };
};
