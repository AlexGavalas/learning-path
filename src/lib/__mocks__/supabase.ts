export const supabase = {
    rpc: jest.fn(),
    upsert: jest.fn(),
    from: jest.fn().mockReturnValue({
        upsert: jest.fn().mockReturnValue({
            error: null,
        }),
    }),
};
