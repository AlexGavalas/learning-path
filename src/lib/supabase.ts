import { createClient } from '@supabase/supabase-js';

import type { Database } from '~types/database.types';

export const supabase = createClient<Database>(
    process.env.PUBLIC_SUPABASE_URL ?? import.meta.env.PUBLIC_SUPABASE_URL,
    process.env.PUBLIC_SUPABASE_ANON_KEY ??
        import.meta.env.PUBLIC_SUPABASE_ANON_KEY,
);
