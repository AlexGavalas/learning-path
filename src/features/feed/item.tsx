import { format } from 'date-fns';
import { type FormEventHandler, useState } from 'react';

import { Button } from '~components/button';
import { Dialog } from '~components/dialog';
import { Textarea } from '~components/textarea';

import { type UserPost } from '../../../types/notes.types';

interface ListItemProps {
    post: UserPost;
    isMine: boolean;
    onPostDelete: (id: string) => void;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const ListItem = ({
    post,
    isMine,
    onPostDelete,
    onPostUpdate,
}: ListItemProps) => {
    const [openEditDialog, setOpenEditDialog] = useState(false);
    const { post: postData, name, id, created_at } = post;

    const postDate = format(new Date(created_at), 'dd/MM/yyyy HH:mm');

    const closeDialog = () => setOpenEditDialog(false);

    const updatePost: FormEventHandler<HTMLFormElement> = async (e) => {
        e.preventDefault();

        const newPost = new FormData(e.currentTarget)
            .get('newpost')
            ?.toString();

        if (!newPost) return;

        await onPostUpdate(post.id, newPost);

        closeDialog();
    };

    return (
        <li className="flex list-none flex-col border-0 border-solid py-6 last:border-b-0">
            <p className="break-words border-0 border-l-2 border-solid border-l-teal-500 pl-4 dark:border-l-yellow-500">
                {postData}
            </p>
            <div className="flex items-center gap-4 self-end">
                {isMine && (
                    <div className="flex h-5 gap-2">
                        <Button
                            variant="danger"
                            onClick={() => onPostDelete(id)}
                        >
                            Delete
                        </Button>
                        <Button
                            onClick={(e) => {
                                e.stopPropagation();
                                setOpenEditDialog(true);
                            }}
                        >
                            Edit
                        </Button>
                    </div>
                )}
                <p className="text-sm italic ">
                    - {name} at {postDate}
                </p>
            </div>
            <Dialog
                open={openEditDialog}
                size="md"
                onClickOutside={closeDialog}
            >
                <form onSubmit={updatePost} className="flex flex-col gap-2">
                    <Textarea
                        label="Make your changes"
                        name="newpost"
                        defaultValue={post.post}
                    />
                    <div className="flex justify-end gap-2">
                        <Button
                            onClick={closeDialog}
                            type="button"
                            variant="danger"
                        >
                            Cancel
                        </Button>
                        <Button>Save</Button>
                    </div>
                </form>
            </Dialog>
        </li>
    );
};
