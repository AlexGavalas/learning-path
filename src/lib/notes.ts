import matter from 'gray-matter';
import { serialize } from 'next-mdx-remote/serialize';
import rehypeAutolinkHeadings from 'rehype-autolink-headings';
import rehypeExternalLinks from 'rehype-external-links';
import rehypeSlug from 'rehype-slug';

import { supabase } from '~lib/supabase';
import { type NoteMDX } from '~types/notes.types';

export const getAllNoteIds = async () => {
    const { data: fileNames, error } = await supabase.storage
        .from('notes_test_1')
        .list();

    if (error) {
        throw error;
    }

    return fileNames.map((file) => {
        return {
            params: {
                id: file.name.replace(/\.mdx$/, ''),
            },
        };
    });
};

export const getNoteData = async (filename: string): Promise<NoteMDX> => {
    const filePath = `${filename}.mdx`;

    const { data: fileContents, error } = await supabase.storage
        .from('notes_test_1')
        .download(filePath);

    if (error) {
        throw error;
    }

    const matterResult = matter(await fileContents.text());

    const mdxSource = await serialize(matterResult.content, {
        mdxOptions: {
            rehypePlugins: [
                [rehypeExternalLinks, { target: '_blank' }],
                [rehypeSlug],
                [rehypeAutolinkHeadings, { behavior: 'wrap' }],
            ],
        },
    });

    // Can't type gray matter result yet
    return {
        mdxSource,
        ...matterResult.data,
    } as NoteMDX;
};
