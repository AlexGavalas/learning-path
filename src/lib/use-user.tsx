import {
    FC,
    ReactNode,
    useEffect,
    useState,
    useContext,
    createContext,
} from 'react';

import { Session, User, AuthChangeEvent } from '@supabase/supabase-js';

import { supabase } from '~lib/supabase';

type UserContextType = {
    user: User | null;
    userLoaded: boolean;
};

export const UserContext = createContext<UserContextType | null>(null);

const handleAuthChange = async (
    event: AuthChangeEvent,
    session: Session | null,
) => {
    await fetch('/api/auth', {
        body: JSON.stringify({ event, session }),
        credentials: 'same-origin',
        headers: new Headers({ 'Content-Type': 'application/json' }),
        method: 'POST',
    });
};

export const UserContextProvider: FC<{ children?: ReactNode }> = ({
    children,
}) => {
    const [userLoaded, setUserLoaded] = useState(true);
    const [user, setUser] = useState<User | null>(null);

    useEffect(() => {
        supabase.auth.getSession().then(({ data: { session } }) => {
            setUser(session?.user ?? null);
            setUserLoaded(true);
        });

        const {
            data: { subscription: authListener },
        } = supabase.auth.onAuthStateChange((event, session) => {
            handleAuthChange(event, session);
            setUser(session?.user ?? null);
        });

        return () => {
            authListener.unsubscribe();
        };
    }, []);

    const value = {
        user,
        userLoaded,
    };

    return (
        <UserContext.Provider value={value}>{children}</UserContext.Provider>
    );
};

export const useUser = () => {
    const context = useContext(UserContext);

    if (!context) {
        throw new Error(`useUser must be used within a UserContextProvider.`);
    }

    return context;
};
