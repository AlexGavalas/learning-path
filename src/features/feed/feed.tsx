import { useState, useCallback } from 'react';

import { useUser } from '~lib/use-user';
import { Dialog } from '~components/dialog';
import { Button } from '~components/button';
import { ListItem } from './item';

interface FeedProps {
    posts: UserPost[];
    onPostDelete: (id: string) => Promise<void>;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const Feed = ({ posts, onPostDelete, onPostUpdate }: FeedProps) => {
    const [openDialog, setOpenDialog] = useState(false);
    const [idToDelete, setIdToDelete] = useState('');
    const [error, setError] = useState<string>();
    const { user } = useUser();

    const closeDialog = useCallback(() => setOpenDialog(false), []);

    const handleDelete = (id: string) => {
        setIdToDelete(id);
        setOpenDialog(true);
    };

    const deleteCurrentId = async () => {
        try {
            await onPostDelete(idToDelete);

            closeDialog();
        } catch (e) {
            const errorMessage =
                e instanceof Error ? e.message : 'An unknown error occured';

            setError(errorMessage);
        }
    };

    if (!posts.length) {
        return <p className="text-center">No posts yet</p>;
    }

    return (
        <>
            <ul className="mt-4">
                {posts.map((post) => {
                    const isMine = post.name === user?.user_metadata.name;

                    return (
                        <ListItem
                            key={post.id}
                            post={post}
                            isMine={isMine}
                            onPostDelete={handleDelete}
                            onPostUpdate={onPostUpdate}
                        />
                    );
                })}
            </ul>
            <Dialog open={openDialog} onClickOutside={closeDialog}>
                <p>Are you sure you want to delete?</p>
                {error && <span>{error}</span>}
                <div className="flex gap-2 justify-end">
                    <Button onClick={closeDialog}>No</Button>
                    <Button onClick={deleteCurrentId}>Yes</Button>
                </div>
            </Dialog>
        </>
    );
};
