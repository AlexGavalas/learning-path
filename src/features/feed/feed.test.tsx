import { axe } from 'jest-axe';

import { supabase } from '~lib/supabase';
import { UserContextProvider } from '~lib/use-user';
import { act, renderWithUser, screen, waitFor } from '~test/helpers';

import { type UserPost } from '../../../types/notes.types';
import { Feed } from './feed';

jest.mock('~lib/supabase');

const POSTS: UserPost[] = [
    {
        id: '1',
        created_at: new Date('2022-01-01').toISOString(),
        name: 'name',
        post: 'post',
    },
];

type GetSessionResponse = Awaited<ReturnType<typeof supabase.auth.getSession>>;

const sessionWithUser: GetSessionResponse = {
    error: null,
    data: {
        session: {
            access_token: '',
            expires_in: 0,
            refresh_token: '',
            token_type: '',
            user: {
                app_metadata: {},
                aud: '',
                created_at: '',
                id: '1',
                user_metadata: {
                    name: 'name',
                },
            },
        },
    },
};

const renderFeed = ({ posts }: { posts: UserPost[] }) => {
    const mockPostDelete = jest.fn();
    const mockPostUpdate = jest.fn();

    const renderResult = renderWithUser(
        <Feed
            posts={posts}
            onPostDelete={mockPostDelete}
            onPostUpdate={mockPostUpdate}
        />,
        { wrapper: UserContextProvider },
    );

    return {
        ...renderResult,
        mockPostDelete,
        mockPostUpdate,
    };
};

describe('<Feed />', () => {
    describe('when posts are empty', () => {
        it('renders', () => {
            const { container } = renderFeed({ posts: [] });

            expect(container).toMatchSnapshot();
        });

        it('is accessible', async () => {
            const { container } = renderFeed({ posts: [] });

            const a11yResults = await axe(container);

            expect(a11yResults).toHaveNoViolations();
        });
    });

    describe('when posts exist', () => {
        it('renders', () => {
            const { container } = renderFeed({ posts: POSTS });

            expect(container).toMatchSnapshot();
        });

        it('is accessible', async () => {
            const { container } = renderFeed({ posts: POSTS });

            const a11yResults = await axe(container);

            expect(a11yResults).toHaveNoViolations();
        });

        describe('when user clicks delete', () => {
            const mockGetSession = jest.mocked(supabase.auth.getSession);

            beforeAll(() => {
                mockGetSession.mockResolvedValue(sessionWithUser);
            });

            afterAll(() => {
                jest.resetAllMocks();
            });

            describe('when onPostDelete resolves', () => {
                it('opens the confirm dialog', async () => {
                    const { user, mockPostDelete } = await act(async () =>
                        renderFeed({ posts: POSTS }),
                    );

                    await user.click(screen.getByText(/delete/i));

                    expect(screen.getByRole('dialog')).toBeInTheDocument();

                    await user.click(screen.getByText(/yes/i));

                    expect(mockPostDelete).toHaveBeenCalledTimes(1);
                    expect(mockPostDelete).toHaveBeenCalledWith(POSTS[0].id);

                    await waitFor(() =>
                        expect(
                            screen.queryByRole('dialog'),
                        ).not.toBeInTheDocument(),
                    );
                });
            });

            describe('when onPostDelete rejects', () => {
                it('shows the error message', async () => {
                    const { user, mockPostDelete } = await act(async () =>
                        renderFeed({ posts: POSTS }),
                    );

                    const errorMessage = 'error';

                    mockPostDelete.mockRejectedValue(new Error(errorMessage));

                    await user.click(screen.getByText(/delete/i));
                    await user.click(screen.getByText(/yes/i));

                    expect(screen.getByRole('dialog')).toBeInTheDocument();
                    expect(screen.getByText(errorMessage)).toBeInTheDocument();
                });

                it('shows a generic error', async () => {
                    const { user, mockPostDelete } = await act(async () =>
                        renderFeed({ posts: POSTS }),
                    );

                    mockPostDelete.mockRejectedValue({});

                    await user.click(screen.getByText(/delete/i));
                    await user.click(screen.getByText(/yes/i));

                    expect(screen.getByRole('dialog')).toBeInTheDocument();
                    expect(
                        screen.getByText(/unknown error/i),
                    ).toBeInTheDocument();
                });
            });
        });
    });
});
