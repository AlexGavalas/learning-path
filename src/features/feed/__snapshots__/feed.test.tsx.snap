// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feed /> when posts are empty renders 1`] = `
<div>
  <p
    class="text-center"
  >
    No posts yet
  </p>
</div>
`;

exports[`<Feed /> when posts exist renders 1`] = `
<div>
  <ul
    class="mt-4"
  >
    <li
      class="flex list-none flex-col border-0 border-solid py-6 last:border-b-0"
    >
      <p
        class="border-l-light-primary dark:border-l-dark-primary break-words border-0 border-l-2 border-solid pl-4"
      >
        post
      </p>
      <div
        class="flex items-center gap-4 self-end"
      >
        <p
          class="text-sm italic "
        >
          - 
          name
           at 
          01/01/2022 00:00
        </p>
      </div>
    </li>
  </ul>
</div>
`;
